authn.py

```diff
@@ -0,0 +1,144 @@
+import os
+from dataclasses import dataclass
+
+from flask import g, request
+from functools import wraps
+from jose import jwt
+
+from typing import Dict, List, Union
+from urllib.parse import urlparse
+
+import requests
+
+
+OidcConfig = Dict[str, Union[str, List[str]]]
+Key = Dict[str, str]
+KeySet = Dict[str, List[Key]]
+
+@dataclass(frozen=True)
+class AuthError(Exception):
+    description: str
+    code: int
+
+def get_token_auth_header():
+    """Obtains the Access Token from the Authorization Header
+    """
+    auth = request.headers.get("Authorization", None)
+    if not auth:
+        raise AuthError("Authorization header is expected", 401)
+
+    parts = auth.split()
+
+    if parts[0].lower() != "bearer":
+        raise AuthError("Authorization header must start with Bearer", 401)
+    elif len(parts) == 1:
+        raise AuthError("Token not found", 401)
+    elif len(parts) > 2:
+        raise AuthError("Authorization header must be Bearer token", 401)
+
+    token = parts[1]
+    return token
+
+def requires_auth(f):
+    """Determines if the Access Token is valid
+    """
+    @wraps(f)
+    def decorated(*args, **kwargs):
+        missing_variables = []
+        oidc_issuer = os.getenv("ISSUER", "")
+        if not oidc_issuer:
+            missing_variables.append("ISSUER")
+
+        oidc_client_id = os.getenv("AUDIENCE", "")
+        if not oidc_client_id:
+            missing_variables.append("AUDIENCE")
+
+        if missing_variables:
+            raise EnvironmentError(f"environment variables not set: {', '.join(missing_variables)}")
+        
+        token = get_token_auth_header()
+        key_id = get_key_id(token)
+        iss_url = issuer_url(oidc_issuer)
+        discovery_url = os.path.join(oidc_issuer, ".well-known/openid-configuration")
+        key = find_signing_key(discovery_url, key_id)
+
+        options = {"verify_at_hash": None}
+        claims = jwt.decode(token, key, options=options, audience=oidc_client_id)
+        if "azp" in claims and claims["azp"] != oidc_client_id:
+            raise AuthError(f"'azp' claim '{claims['azp']}' does not match client ID", 401)
+
+        if not isinstance(claims["sub"], str):
+            raise AuthError(f"'sub' claim '{claims['sub']}'is not a valid identity", 401)
+        
+
+        g.identity = claims["sub"]
+        return f(*args, **kwargs)
+    return decorated
+
+
+def get_key_id(token: str) -> str:
+    kid = jwt.get_unverified_header(token).get("kid")
+    if not kid:
+        raise AuthError("Bearer token does not have 'kid' claim", 401)
+
+    return kid  # type: ignore
+
+def find_signing_key(discovery_url: str, key_id: str) -> Key:
```

Comment:
As an optimization, you can decorate this function with `@lru_cache` (see https://docs.python.org/3.8/library/functools.html#functools.lru_cache).
That would cache return values so that subsequent calls with the same url and key_id would not have to to download the signing key again.